### PR TITLE
[ruler] Snap next point to the closest bookmark

### LINKED ISSUE #5966: Snap ruler to user marks
**A problem:**
1. Start building ruler route
2. Tap on a user mark

Actual result: Ruler route is continued to some point near user mark

**Solution:**
Best option is to snap ruler point to the user mark.

### PR DESCRIPTION
**Pull request by strump**
If there's bookmark near tap position then continue route to that bookmark:

| Before | After |
| ---- | ---- |
| <video src="https://github.com/organicmaps/organicmaps/assets/720808/dd003406-7178-41be-8cd4-39a2a3929a73" /> | <video src="https://github.com/organicmaps/organicmaps/assets/720808/d82136d3-c3b5-4575-b03f-14e14bcf02e7" /> |

Closes #5966

### CODE REVIEW

File: map/framework.cpp [hunk 1]
```diff
@@ -2059,7 +2059,12 @@ void Framework::OnTapEvent(place_page::BuildInfo const & buildInfo)
     data.m_pointType = RouteMarkType::Finish;
     data.m_intermediateIndex = m_routingManager.GetRoutePointsCount();
     data.m_isMyPosition = false;
-    data.m_position = buildInfo.m_mercator;
+
+    if (placePageInfo.has_value() && placePageInfo->IsBookmark())
```
Comment (vng) [inline map/framework.cpp:2063]: nit: if (placePageInfo && ...

**Review by vng — APPROVED** [review]
LGTM, but it is a but fancy way. This logic expects that PP was shown beforehand ...
Better to fetch bookmarks from container and snap by coordinates.

**Review by biodranik — COMMENTED** [review]
Should Framework::OnTapEvent be called at all when users are building the route with a ruler? The PP is not shown. It may break in the future implementation.

You can check the callstack and find how buildinfo was constructed earlier. That would be the right place to check bookmarks containers.

**Comment by strump** [comment]
@biodranik  I debugged tap event handler and found that initially method `void FrontendRenderer::OnTap(m2::PointD const & pt, bool isLongTap)` is called (see [frontend_renderer.cpp:1978](https://github.com/organicmaps/organicmaps/blob/master/drape_frontend/frontend_renderer.cpp#L1978)). This method looks for POI, bookmark or my position and invokes `Framework::OnTapEvent(...)`.

There is no `RoutingManager` in `FrontendRenderer` to check if we are in ruler mode. Should I modify `frontend_renderer.cpp` by adding ruler mode flag and ruler tap handler? What would be the best place for ruler tap logic?

File: map/framework.cpp (hunk 1 above)
Comment (strump) [reply map/framework.cpp]: fixed

**Review by biodranik — APPROVED** [review]

**Pull request merged**